eth/core/precompile/factories.go

```diff
@@ -92,61 +97,90 @@ func (sf *StatefulFactory) Build(
-	if precompileMethods := sci.PrecompileMethods(); precompileMethods != nil {
-		idsToMethods, err = sf.buildIdsToMethods(precompileMethods, sci.ABIMethods())
-		if err != nil {
-			return nil, err
-		}
+	idsToMethods, err := buildIdsToMethods(sci.ABIMethods(), reflect.ValueOf(sci))
+	if err != nil {
+		return nil, err
 	}
 
-	return NewStateful(rp, idsToMethods), nil
+	return NewStateful(rp, idsToMethods)
 }
 
-// buildIdsToMethods builds the stateful precompile container for the given `precompileMethods`
-// and `abiMethods`. This function will return an error if every method in `abiMethods` does not
-// have a valid, corresponding `Method`.
-func (sf *StatefulFactory) buildIdsToMethods(
-	precompileMethods Methods,
-	abiMethods map[string]abi.Method,
-) (map[string]*Method, error) {
-	// validate precompile methods
-	for _, pm := range precompileMethods {
-		if err := pm.ValidateBasic(); err != nil {
-			return nil, err
-		}
-	}
-
-	// match every ABI method to corresponding precompile method
+// This function matches each Go implementation of the precompile to the ABI's respective function.
+// It searches for the ABI function in the Go precompile contract and performs basic validation on
+// the implemented function.
+func buildIdsToMethods(pcABI map[string]abi.Method, contractImpl reflect.Value) (map[string]*Method, error) {
+	contractImplType := contractImpl.Type()
 	idsToMethods := make(map[string]*Method)
-	for name := range abiMethods {
-		abiMethod := abiMethods[name]
-
-		// find the corresponding precompile method for abiMethod based on signature
-		var precompileMethod *Method
-		i := 0
-		for ; i < len(precompileMethods); i++ {
-			if precompileMethods[i].AbiSig == abiMethod.Sig {
-				precompileMethod = precompileMethods[i]
-				break
+	for m := 0; m < contractImplType.NumMethod(); m++ {
+		implMethod := contractImplType.Method(m)
+		implMethodName := formatName(implMethod.Name)
+
+		if abiMethod, found := pcABI[implMethodName]; found {
+			if err := checkReturnTypes(implMethod); err != nil {
+				return nil, errorslib.Wrap(err, implMethodName)
 			}
+			idsToMethods[utils.UnsafeBytesToStr(abiMethod.ID)] = NewMethod(
+				&abiMethod,
+				abiMethod.Sig,
+				implMethod.Func,
+			)
 		}
-		if i == len(precompileMethods) {
-			return nil, errors.Wrap(ErrNoPrecompileMethodForABIMethod, abiMethod.Sig)
-		}
+	}
 
-		// attach the ABI method to the precompile method for stateful container to handle
-		precompileMethod.AbiMethod = &abiMethod
-		idsToMethods[utils.UnsafeBytesToStr(abiMethod.ID)] = precompileMethod
+	for _, abiMethod := range pcABI {
+		if _, found := idsToMethods[utils.UnsafeBytesToStr(abiMethod.ID)]; !found {
+			return nil, errorslib.Wrap(ErrNoPrecompileMethodForABIMethod, abiMethod.Name)
+		}
 	}
+
 	return idsToMethods, nil
 }
+
+// formatName converts to first character of name to lowercase. If the first
+// three characters are "ERC" or "ABI" (which is p common), then it converts all
```

Comment:
this is probably a bad idea lmfaoooo, i can already see the gremlins this is gonna cause.

Reply:
thoughts on making the go function name exactly match (and subsequently unexport them) the solidity funcion names then? this was more of an edge case for functions like `ERC20AddressForCoinDenom` in `erc20.go`

Reply:
i think just make it be weird tbh at least it's consistent